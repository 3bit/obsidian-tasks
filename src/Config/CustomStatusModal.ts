import { Modal, Setting, TextComponent } from 'obsidian';
import { Status, StatusConfiguration } from '../Status';
import type TasksPlugin from '../main';

const isInvalidClass = 'is-invalid';
const hasInvalidMessage = 'has-invalid-message';

const unsetAlignItems = 'unset-align-items';
const dotUnsetAlignItems = '.' + unsetAlignItems;

const invalidFeedback = 'invalid-feedback';
const dotInvalidFeedback = '.' + 'invalid-feedback';

export class CustomStatusModal extends Modal {
    statusSymbol: string;
    statusName: string;
    statusNextSymbol: string;
    statusAvailableAsCommand: boolean;
    saved: boolean = false;
    error: boolean = false;
    constructor(public plugin: TasksPlugin, statusType: StatusConfiguration) {
        super(plugin.app);
        this.statusSymbol = statusType.indicator;
        this.statusName = statusType.name;
        this.statusNextSymbol = statusType.nextStatusIndicator;
        this.statusAvailableAsCommand = statusType.availableAsCommand;
    }

    /**
     * Return a {@link StatusConfiguration} from the modal's contents
     */
    public statusConfiguration() {
        return new StatusConfiguration(
            this.statusSymbol,
            this.statusName,
            this.statusNextSymbol,
            this.statusAvailableAsCommand,
        );
    }

    async display() {
        const { contentEl } = this;

        contentEl.empty();

        const settingDiv = contentEl.createDiv();
        //const title = this.title ?? '...';

        let statusSymbolText: TextComponent;
        new Setting(settingDiv)
            .setName('Task Status Symbol')
            .setDesc('This is the character between the square braces.')
            .addText((text) => {
                statusSymbolText = text;
                statusSymbolText.setValue(this.statusSymbol).onChange((v) => {
                    if (!v.length) {
                        CustomStatusModal.setValidationError(text, 'Task status type cannot be empty.');
                        return;
                    }

                    if (v.includes(' ')) {
                        CustomStatusModal.setValidationError(text, 'Task status type cannot include spaces.');
                        return;
                    }

                    if (v.length > 1) {
                        CustomStatusModal.setValidationError(text, 'Task status must be a single character.');
                        return;
                    }
                    CustomStatusModal.removeValidationError(text);
                    this.statusSymbol = v;
                });
            });

        new Setting(settingDiv)
            .setName('Task Status Name')
            .setDesc('This is the friendly name of the task status.')
            .addText((text) => {
                text.setValue(this.statusName).onChange((v) => {
                    this.statusName = v;
                });
            });

        new Setting(settingDiv)
            .setName('Task Next Status Symbol')
            .setDesc('When clicked on this is the symbol that should be used next.')
            .addText((text) => {
                text.setValue(this.statusNextSymbol).onChange((v) => {
                    this.statusNextSymbol = v;
                });
            });

        if (Status.tasksPluginCanCreateCommandsForStatuses()) {
            new Setting(settingDiv)
                .setName('Available as command')
                .setDesc(
                    'If enabled this status will be available as a command so you can assign a hotkey and toggle the status using it.',
                )
                .addToggle((toggle) => {
                    toggle.setValue(this.statusAvailableAsCommand).onChange(async (value) => {
                        this.statusAvailableAsCommand = value;
                    });
                });
        }

        const footerEl = contentEl.createDiv();
        const footerButtons = new Setting(footerEl);
        footerButtons.addButton((b) => {
            b.setTooltip('Save')
                .setIcon('checkmark')
                .onClick(async () => {
                    // let error = false;
                    // if (!this.statusSymbol.length) {
                    //     SettingsModal.setValidationError(this.statusSymbol, 'Task status type cannot be empty.');
                    //     error = true;
                    //     return;
                    // }

                    // if (error) {
                    //     new Notice('Fix errors before saving.');
                    //     return;
                    // }
                    this.saved = true;
                    this.close();
                });
            return b;
        });
        footerButtons.addExtraButton((b) => {
            b.setIcon('cross')
                .setTooltip('Cancel')
                .onClick(() => {
                    this.saved = false;
                    this.close();
                });
            return b;
        });
    }

    // updateTitle(admonitionPreview: HTMLElement, title: string) {
    //     let titleSpan = admonitionPreview.querySelector('.admonition-title-content');
    //     let iconEl = admonitionPreview.querySelector('.admonition-title-icon');
    //     titleSpan.textContent = title;
    //     titleSpan.prepend(iconEl);
    // }
    onOpen() {
        this.display();
    }

    static setValidationError(textInput: TextComponent, message?: string) {
        textInput.inputEl.addClass(isInvalidClass);
        if (message) {
            textInput.inputEl.parentElement?.addClasses([hasInvalidMessage, unsetAlignItems]);
            textInput.inputEl.parentElement?.parentElement?.addClass(dotUnsetAlignItems);
            let mDiv = textInput.inputEl.parentElement?.querySelector(dotInvalidFeedback) as HTMLDivElement;

            if (!mDiv) {
                mDiv = createDiv({ cls: invalidFeedback });
            }
            mDiv.innerText = message;
            mDiv.insertAfter(textInput.inputEl);
        }
    }
    static removeValidationError(textInput: TextComponent) {
        textInput.inputEl.removeClass(isInvalidClass);
        textInput.inputEl.parentElement?.removeClasses([hasInvalidMessage, unsetAlignItems]);
        textInput.inputEl.parentElement?.parentElement?.removeClass(dotUnsetAlignItems);

        const invalidFeedback = textInput.inputEl.parentElement?.querySelector(dotInvalidFeedback);
        if (invalidFeedback) {
            textInput.inputEl.parentElement?.removeChild(invalidFeedback);
        }
    }
}
