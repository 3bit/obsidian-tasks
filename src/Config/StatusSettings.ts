import type { StatusConfiguration } from '../Status';

export class StatusSettings {
    constructor() {
        this.customStatusTypes = [];
    }
    customStatusTypes: StatusConfiguration[];
}
