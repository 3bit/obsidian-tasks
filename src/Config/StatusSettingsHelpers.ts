/* This file contains re-usable functions for use in managing and editing
 * user settings for custom task statuses.
 * It intentionally does not import any Obsidian types, so that tests can
 * be written for its contents.
 */
import { StatusConfiguration } from '../Status';

/**
 * Add a collection of of supported statuses to an existing collection of StatusConfiguration objects.
 * This can be used to quickly populate the user's settings.
 * If there are any exact duplicates already present, they are skipped, and noted in the returned value.
 *
 * @param supportedStatuses - an array of status specifications, for example `['b', 'Bookmark', 'x']`
 * @param statusTypes {@link StatusConfiguration} - an array of existing known statuses
 * @return An array of warning messages to show the user, one for each rejected exact duplicate status.
 *
 */
export function addCustomStatusesCollection(
    supportedStatuses: Array<[string, string, string]>,
    statusTypes: StatusConfiguration[],
): string[] {
    const notices: string[] = [];
    supportedStatuses.forEach((importedStatus) => {
        const hasStatus = statusTypes.find((element) => {
            return (
                element.indicator == importedStatus[0] &&
                element.name == importedStatus[1] &&
                element.nextStatusIndicator == importedStatus[2]
            );
        });
        if (!hasStatus) {
            statusTypes.push(new StatusConfiguration(importedStatus[0], importedStatus[1], importedStatus[2], false));
        } else {
            notices.push(`The status ${importedStatus[1]} (${importedStatus[0]}) is already added.`);
        }
    });
    return notices;
}
