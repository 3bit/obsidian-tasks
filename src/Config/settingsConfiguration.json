[
  {
    "text": "Placeholder for Tasks Status Types",
    "level": "h1",
    "class": "",
    "open": true,
    "notice": {
      "class": "",
      "text": "This will end up allowing customisation of task statuses.",
      "html": ""
    },
    "settings": []
  }
]
