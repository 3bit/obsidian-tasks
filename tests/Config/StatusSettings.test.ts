import { verifyAsJson } from 'approvals/lib/Providers/Jest/JestApprovals';
import { StatusSettings } from '../../src/Config/StatusSettings';

describe('StatusSettings', () => {
    it('verify default status settings', () => {
        const defaultStatusSettings = new StatusSettings();
        verifyAsJson(defaultStatusSettings);
    });
});
