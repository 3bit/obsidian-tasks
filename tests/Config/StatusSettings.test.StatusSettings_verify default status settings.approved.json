{
  "coreStatusTypes": [
    {
      "symbol": " ",
      "name": "Todo",
      "nextStatusSymbol": "x",
      "availableAsCommand": true,
      "type": "TODO"
    },
    {
      "symbol": "/",
      "name": "In Progress",
      "nextStatusSymbol": "x",
      "availableAsCommand": true,
      "type": "IN_PROGRESS"
    },
    {
      "symbol": "x",
      "name": "Done",
      "nextStatusSymbol": " ",
      "availableAsCommand": true,
      "type": "DONE"
    },
    {
      "symbol": "-",
      "name": "Cancelled",
      "nextStatusSymbol": " ",
      "availableAsCommand": true,
      "type": "CANCELLED"
    }
  ],
  "customStatusTypes": []
}
